### PR TITLE
No-require rule 

### PR DESCRIPTION
**Pull request by neriyarden**

### CODE REVIEW

File: src/rules/no-commonjs.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
+/**
+ * @fileoverview Rule to prefer ES6 to CJS
+ * @author Jamund Ferguson
```
Comment (unicop) [inline src/rules/no-commonjs.ts:3]: I guess you can replace the author name

File: src/rules/no-commonjs.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+function validateScope(scope) {
+    return scope.variableScope.type === "module";
+}
+
+// https://github.com/estree/estree/blob/HEAD/es5.md
+function isConditional(node) {
+    if (
+        node.type === "IfStatement" ||
```
Comment (unicop) [inline src/rules/no-commonjs.ts:45]: Please use the `AST_NODE_TYPES` instead of literal strings

File: src/rules/no-commonjs.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    type: "object",
+    properties: {
+        allowPrimitiveModules: { type: "boolean" },
+        allowRequire: { type: "boolean" },
+        allowConditionalRequire: { type: "boolean" },
+    },
+    additionalProperties: false,
+};
+
+export default createRule({
```
Comment (unicop) [inline src/rules/no-commonjs.ts:77]: I will publish how we should work with the types system typescript-eslint has implemented

File: src/rules/no-commonjs.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import { createRule } from "../util";
+import docsUrl from "../docsUrl";
+type MessageIds = "importMessage" | "exportMesage";
+
+type Options =
+    | ["allow-primitive-modules"]
+    | [{ allowPrimitiveModules: boolean }];
+type Context = Readonly<TSESLint.RuleContext<MessageIds, Options>>;
+
+function normalizeLegacyOptions(options: Options) {
```
Comment (unicop) [inline src/rules/no-commonjs.ts:17]: where did you bring this function from?

File: src/rules/no-commonjs.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        const options = normalizeLegacyOptions(context.options);
+
+        return {
+            MemberExpression(node: TSESTree.MemberExpression) {
+                // module.exports
+                if (
+                    node.object.name === "module" &&
+                    node.property.name === "exports"
+                ) {
+                    if (allowPrimitive(node, options)) return;
```
Comment (unicop) [inline src/rules/no-commonjs.ts:119]: primite is like returning a literal string from a module, e.g. `module.exports = 'bla'` ?

File: src/rules/no-commonjs.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                        .variables.some(
+                            (variable) => variable.name === "exports"
+                        );
+                    if (!isInScope) {
+                        context.report({ node, message: "exportMessage" });
+                    }
+                }
+            },
+            CallExpression: function (call) {
+                if (!validateScope(context.getScope())) return;
```
Comment (unicop) [inline src/rules/no-commonjs.ts:136]: not sure what is this part
Did you implement a general no commonjs require/exports?
This should only check that `module.exports` is not exists
Although I like the scope, to verify we are at a module file

**Review by unicop — COMMENTED** [review]
@neri left comments 🙃

File: src/rules/no-commonjs.ts (hunk 1 above)
Comment (neriyarden) [reply src/rules/no-commonjs.ts:3]: i wouldn't take full credit here.. i didn't write it all by myself :)

File: src/rules/no-commonjs.ts (hunk 3 above)
Comment (neriyarden) [reply src/rules/no-commonjs.ts:65]: cool, good idea.
also, i added some types now

File: src/rules/no-commonjs.ts (hunk 4 above)
Comment (neriyarden) [reply src/rules/no-commonjs.ts]: it's from the `eslint-import-plugin` rule.
i deleted it though as it's confusing.

I think we can stick to 1 way of providing the options (in an object)

File: src/rules/no-commonjs.ts (hunk 6 above)
Comment (neriyarden) [reply src/rules/no-commonjs.ts:118]: im not sure either.. it was in the original rule and i didn't figure it out yet.

what about the `exports` section?we want to keep it?

File: src/rules/no-commonjs.ts (hunk 5 above)
Comment (neriyarden) [reply src/rules/no-commonjs.ts:101]: yes.
we don't need it. im removing it

**Review by unicop — APPROVED** [review]
Amazing job here!
we are on the right track for creating impact 🥇

File: src/rules/no-require.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                additionalProperties: false,
+            },
+        ],
+    },
+
+    create(context) {
+        const { allowConditionalRequire = false } = context.options[0] || {};
+
+        return {
+            CallExpression: function (call) {
```
[the remark this answers is not in the document]
Comment (neriyarden) [reply src/rules/no-require.ts]: hmm im not sure.. i think maybe calling all of the node types `node` is less confusing

File: tests/rules/no-require.test.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+});
+
+ruleTester.run('no-require', rule, {
+    valid: [
+        // imports
+        { code: 'import "x";' },
+        { code: 'import x from "x"' },
+        { code: 'import { x } from "x"' },
+
+        // allowed requires - which ones do we want to allow?
```
[the remark this answers is not in the document]
Comment (neriyarden) [reply tests/rules/no-require.test.ts]: this is an old comment 🙃

File: tests/rules/no-require.test.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+/* eslint "@typescript-eslint/internal/plugin-test-formatting": ["error", { formatWithPrettier: false }] */
+/* eslint-enable eslint-comments/no-use */
+import {
+    InvalidTestCase,
+    ValidTestCase,
+} from '@typescript-eslint/utils/dist/ts-eslint';
+import rule, * as noRequireRule from '../../src/rules/no-require';
+import { RuleTester } from '../RuleTester';
+
+const ruleTester = new RuleTester({
```
[the remark this answers is not in the document]
Comment (neriyarden) [reply tests/rules/no-require.test.ts:12]: it is :)
so convenient..

**Review by unicop — APPROVED** [review]
Merging tomorrow morning

**Pull request merged**